conda_build/build.py

```diff
@@ -1802,10 +1809,26 @@ def bundle_conda(output, metadata: MetaData, env, stats, **kw):
     if files:
         # Files is specified by the output
         # we exclude the list of files that we want to keep, so post-process picks them up as "new"
-        keep_files = {
-            os.path.normpath(pth)
-            for pth in utils.expand_globs(files, metadata.config.host_prefix)
-        }
+        if isinstance(files, dict):
+            # When file matching with include/exclude lists, only
+            # new_prefix_files are considered. Files in the PREFIX from other
+            # recipes (dependencies) are ignored
+            include = files.get("include") or []
+            exclude = files.get("exclude") or []
+            exclude_files = {
+                os.path.normpath(pth)
+                for pth in utils.expand_globs(exclude, metadata.config.host_prefix)
+            }
+            keep_files = {
+                os.path.normpath(pth)
+                for pth in utils.expand_globs(include, metadata.config.host_prefix)
+            }
+            keep_files = new_prefix_files.intersection(keep_files) - exclude_files
+        else:
+            keep_files = {
+                os.path.normpath(pth)
+                for pth in utils.expand_globs(files, metadata.config.host_prefix)
+            }
```

Comment:
Given the change in behavior above with only allowing new files to be included should this behavior here be deprecated?

```suggestion
        else:
            keep_files = {
                os.path.normpath(pth)
                for pth in utils.expand_globs(files, metadata.config.host_prefix)
            }
            
            if keep_files - new_prefix_files:
                deprecated.topic(...)
```

Reply:
It's still useful for eg when building `gcc_bootstrap` packages. Deprecating the implicit behaviour and adding a new flag would be useful.

Reply:
I feel that the current behavior of including all files from the prefix for multi-output and only installed files from the prefix for single-output recipes is inconsistent and thus non-intuitive. [rattler-build is adding a section called "always_include_files"](https://github.com/prefix-dev/rattler-build/pull/819#issuecomment-2148525698) for recipes that want to repackage artifacts from their host dependencies.

I'm favor of deprecating the old behavior, but I'm uncertain about any removal or replacement schedule.

Reply:
> [rattler-build is adding a section called "always_include_files"](https://github.com/prefix-dev/rattler-build/pull/819#issuecomment-2148525698) for recipes that want to repackage artifacts from their host dependencies.

This exists in conda already, and is [in use](https://github.com/search?q=org%3Aconda-forge%20%2Falways_include_files%3A%2F%20path%3Arecipe%2Fmeta.yaml&type=code) in conda-forge. Main use-case for that from my POV is to be explicit about overwriting some CMake metadata if an incremental build adds more targets to an already-existing CMake file ([example](https://github.com/conda-forge/abseil-cpp-feedstock/issues/62)).

Reply:
We can deprecate that behaviour in a separate PR too. This PR has been open for too long and I wouldn't want to test your patience further @carterbox <3

Reply:
I'll open an issue to capture this feedback, though I think with the documentation this PR added we are in a better place already.